bundler/lib/dependabot/bundler/update_checker/latest_version_finder/dependency_source.rb

```diff
@@ -0,0 +1,149 @@
+# frozen_string_literal: true
+
+module Dependabot
+  module Bundler
+    class UpdateChecker
+      class LatestVersionFinder
+        class DependencySource
+          require_relative "../shared_bundler_helpers"
+          include SharedBundlerHelpers
+
+          RUBYGEMS = "rubygems"
+          PRIVATE_REGISTRY = "private"
+          GIT = "git"
+          OTHER = "other"
+
+          attr_reader :dependency, :dependency_files, :repo_contents_path,
+                      :credentials
+
+          def initialize(dependency:,
+                         dependency_files:,
+                         repo_contents_path: nil,
+                         credentials:)
+            @dependency          = dependency
+            @dependency_files    = dependency_files
+            @repo_contents_path  = repo_contents_path
+            @credentials         = credentials
+          end
+
+          def versions
+            return rubygems_versions if dependency.name == "bundler"
+            return rubygems_versions unless gemfile
+
+            case source_type
+            when OTHER
+              []
+            when GIT
+              latest_git_version_details
```

Comment:
As noted in my comment above, this method should actually return empty for `git` to keep the contract a reflection of `latest_git_version_details` - the methods should be mutually exclusive.

I do think there's a better way though, so will have a thinking